(WIP)chore(camera): refactor video to use data output buffer

timothyrchew opened the pull request:
Alright, everything seems to be ironed out. Marked `WIP` because I still need to finish removing some legacy methods/vars, but its ready for your eyes to see if there are any issues with this implementation.

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
wip

Ignigena commented on ios/RN/RNCamera.m:
```diff
@@ -423,33 +414,44 @@ - (void)record:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve r
+                    self.timer = [NSTimer scheduledTimerWithTimeInterval:maxDuration target:self selector:@selector(stopAssetWriter) userInfo:nil repeats:NO];
+                    return;
+                }
+                self.timer = [NSTimer scheduledTimerWithTimeInterval:3.0 target:self selector:@selector(stopAssetWriter) userInfo:nil repeats:NO];
```
Rather than scheduling an arbitrary timer length if there is no `maxDuration` can we instead bridge the `stop` action and just require that to be called manually?

Ignigena commented on ios/RN/RNCamera.m:
```diff
@@ -423,33 +414,44 @@ - (void)record:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve r
-        NSURL *outputURL = [[NSURL alloc] initFileURLWithPath:path];
-        [self.movieFileOutput startRecordingToOutputFileURL:outputURL recordingDelegate:self];
+
+        if (self.videoWriter.status != AVAssetWriterStatusWriting) {
```
Does this check need to happen inside of the async dispatch? Or can we check this before even creating the thread? If so, may be worth doing something similar with the timer so we're not dispatching a background thread that dispatches a foreground thread :)

Ignigena commented on ios/RN/RNCamera.h:
```diff
@@ -11,7 +11,7 @@
 @class RNCamera;
 
-@interface RNCamera : UIView <AVCaptureMetadataOutputObjectsDelegate, AVCaptureFileOutputRecordingDelegate, RNFaceDetectorDelegate>
+@interface RNCamera : UIView <AVCaptureMetadataOutputObjectsDelegate, AVCaptureFileOutputRecordingDelegate, RNFaceDetectorDelegate, AVCaptureVideoDataOutputSampleBufferDelegate>
```
As part of the cleanup here, if we aren't using the file output at all, let's remove any orphaned functionality including `AVCaptureFileOutputRecordingDelegate`

Ignigena commented on ios/RN/RNCamera.m:
```diff
@@ -423,33 +414,44 @@ - (void)record:(NSDictionary *)options resolve:(RCTPromiseResolveBlock)resolve r
     });
 }
 
+- (void)stopAssetWriter {
```
To go along with the commend above, might be nice to modify the existing bridge `RNCameraManager.m:250` with a `RCTPromiseResolveBlock` and `RCTPromiseRejectBlock` so the video recording can be used without a timer for flexibility.

Ignigena commented on ios/RN/RNCamera.m:
```diff
@@ -821,7 +836,8 @@ - (void)processVideoToAnimation:(NSURL *)outputFileURL {
 
-    for (int i = 0; i < framesNeeded; i++) {
+    // Start int at 1 to skip first frame timestamped 0.00
+    for (int i = 1; i < framesNeeded; i++) {
```
This will change the animation style from how it is currently. May not be a big deal since it's only dropping two frames. But I would be curious if there's another workaround here. If `0.00` poses a problem, can we just shift the first frame (or all the frames) ever so slightly to `0.01`?

Ignigena reviewed (commented):
This is looking fantastic! I love to see how little code changes are required here. A few bits of feedback but let's talk this over tomorrow (and I want to hear how you fixed the issue with orientation!)

As a sidenote: from what you've been digging into, are there clear benefits to using AVAssetWriter the way we are doing even without the exposure point requirement that has forced us to do so? Would this be worth trying to merge upstream into the parent project particularly since AVCaptureMovieFileOutput and Google Vision are not compatible (see the workaround in place on `RNCamera.m:481`)

Commit:
wip

Commit:
wip

Commit:
cleanup

timothyrchew commented:
@Ignigena Alright, this is ready for another look. Addressed issues above, added some additional AVAssetWriter cleanup and error catching after stream is stopped, and removed old AVCaptureMovieFileOutput related code.
